### PR TITLE
feat(UI & API): Manage views in the frontend

### LINKED ISSUE #336: [UI] Manage views
Relates to #334 

When a project is created, the space where views are presented is empty, or a new project has a view initially created with the name "Default".

**User flow:**
- The user selects the path where to create a view
- A prompt for a name appears
- The empty view is displayed
- The user adds/removes items to/from the view

The user can see all their views in the list, and select another one, updating the currently displayed view.

The user can delete a view they no longer need.

### PR DESCRIPTION
**Pull request by rouk1**
### Backend

All project endpoints are now below the same prefix.

- `GET /api/project/items` lists all items of a project.
- `POST /api/project/views/share/{key}` returns a static shareable HTML file named `{view name}-{yyyy-MM-dd-HH-mm}.html`.
- `PUT /api/project/views/{key}` save a view in the project.
- `DELETE /api/project/views/{key}` delete a view from the project.

### Frontend

- A new `EditableList` component as been implemented to manage views. 
- A lot of renaming to get rid of the concept of "report".
- Plug everything to the new layout.
- Rework view card layout and add an action dropdown to it.
- Rework dropdowns using floating UI to make them fixed.

https://github.com/user-attachments/assets/3f94e483-ecea-4d5a-8433-50f9150fc6e6

Co-authored with @augustebaum.
Fixes #336, #407, #377, #332

### CODE REVIEW

File: frontend/src/services/api.ts [hunk 1]
```diff
@@ -34,11 +34,11 @@ export async function fetchReport(): Promise<Report | null> {
   return null;
 }
 
-export async function putLayout(payload: Layout): Promise<Report | null> {
+export async function putView(view: string, layout: Layout): Promise<Project | null> {
   try {
-    const r = await fetch(`${BASE_URL}/report/layout`, {
+    const r = await fetch(`${BASE_URL}/project/views/${view}`, {
```
Comment (tuscland) [inline frontend/src/services/api.ts:39]: I am not at ease with using names as identifiers in URLs. It often clashes with the `/` character found in the URL path. Could you use a query parameter instead?

For example `/project/views?name=${view}`

Same suggestion for delete and share features.

**Comment by augustebaum** [comment]
Really awesome!

File: frontend/src/views/ProjectView.vue [hunk 2]
```diff
@@ -1,29 +1,33 @@
...
+import { useToastsStore } from "@/stores/toasts";
 
-const reportStore = useReportStore();
+const projectStore = useProjectStore();
+const toastsStore = useToastsStore();
 const isDropIndicatorVisible = ref(false);
 const editor = ref<HTMLDivElement>();
 const isInFocusMode = ref(false);
+const views = ref<EditableListItemModel[]>([]);
+let unsavedViewsIds: string[] = [];
```
Comment (tuscland) [inline frontend/src/views/ProjectView.vue:22]: question: Is this used to manage delayed view saving? I am not sure to know why a view would not be saved.

File: frontend/src/services/api.ts (hunk 1 above)
Comment (augustebaum) [reply frontend/src/services/api.ts:39]: Out of curiosity, what bothers you with this?

Comment (tuscland) [reply frontend/src/services/api.ts:39]: If you include an identifier in the URL, you want a string over which you have control, that is, something that the program created. Here it is a user-supplied name, hence we have not a lot of control over it, unless we strictly apply validation rules.

File: frontend/src/components/EditableListItem.vue [hunk 3]
```diff
@@ -1,24 +1,53 @@
...
     <div class="label-container">
-      <span class="icon" v-if="props.item.icon" :class="props.item.icon" />
-      <span class="label">
-        {{ props.item.name }}
+      <span class="icon" v-if="item.icon" :class="item.icon" />
+      <span
+        class="label"
+        :contenteditable="item.isUnnamed"
+        ref="label"
+        @blur="onItemNameEdited"
```
Comment (augustebaum) [inline frontend/src/components/EditableListItem.vue]: What does `blur` do? I can't find a reference to it in the Vue docs

File: frontend/src/components/DropdownButton.vue
Comment (augustebaum) [inline frontend/src/components/DropdownButton.vue:1]: can you explain why this necessary?

**Review by augustebaum — APPROVED** [review]

**Comment by anasstarfaoui** [comment]
That's an amazing app so far guys, big kuddos to all of you :) 

Just one suggestion, would it be possible to fixed the position of the Views list by default? I feel like it would be preferrable to not have something who grows @rouk1 :)

File: frontend/src/views/ProjectView.vue (hunk 2 above)
Comment (rouk1) [reply frontend/src/views/ProjectView.vue:22]: It is used to differentiate views who are not know by the backend. It's a mapping of pending views. For example you create a view and did not gave it a name then it's local id is in this array and the project view knows that it's a creation and not a rename.

File: frontend/src/components/EditableListItem.vue (hunk 3 above)
Comment (rouk1) [reply frontend/src/components/EditableListItem.vue]: It's a native event of inputs (lose of focus).
https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event

File: frontend/src/services/api.ts (hunk 1 above)
Comment (tuscland) [reply frontend/src/services/api.ts:39]: What about using a proper id?

**Comment by rouk1** [comment]
> Just one suggestion, would it be possible to fixed the position of the Views list by default?

Could you elaborate please I don't understand your comment ? The view list is on top of the page and has a height of 20% of the height of the viewport.

**Comment by thomass-dev** [comment]
Huge work 👏🏻  , the video is awesome. 
It's a bit hard for me to evaluate anything other than python, as the diff is a bit long 😄 .
Will do my best.

File: frontend/src/components/DropdownButton.vue [hunk 4]
```diff
@@ -14,19 +16,63 @@ const props = withDefaults(defineProps<DropdownProps>(), { isPrimary: false, ali
...
+});
 
 function closeDropdown(e: Event) {
   if (el.value && !el.value.contains(e.target as Node)) {
     isOpen.value = false;
   }
 }
 
+// Intersection observer to close the dropdown when it is not visible
+const intersectionObserver = new IntersectionObserver(
```
[the remark this answers is not in the document]
Comment (rouk1) [reply frontend/src/components/DropdownButton.vue]: Dropdown were not used in scrolling area we did not have that need before...

File: frontend/src/services/api.ts (hunk 1 above)
Comment (rouk1) [reply frontend/src/services/api.ts:39]: I suggest that we use querystring for now and generalize the usage of id's in another issue.
WDYT ?

**Review by augustebaum — CHANGES_REQUESTED** [review]
I found a UX issue: when first starting the UI, adding cards to the canvas doesn't work, with no explanation why. I think it's because no view is selected by default.
Some ways to solve this:
- Select the first view in the list at startup
- Allow the user to set a "default" view, and select that one at startup
- Warn the user that they need to select a view

The easiest is probably the warning. @anasstarfaoui do you see a better way of dealing with this?

EDIT: Resolved

**Review by augustebaum — COMMENTED** [review]
I think the "intermediate view" mechanic when creating a view is a bit strange. I would expect a view either to exist, or not to exist. Some alternatives we discussed with @rouk1:
- If the user clicks away during the view naming process, either validate or cancel the view creation (like in VSCode when you create a file: if you click away and the text box is empty, the creation is aborted; if the text box is not empty, the file is created). Essentially, clicking away should be as effectful as pressing "Enter"
- Ask for the view name before creating anything (e.g. in a modal)

File: frontend/src/services/api.ts (hunk 1 above)
Comment (rouk1) [reply frontend/src/services/api.ts]: Query params are now used.

**Comment by rouk1** [comment]
> I found a UX issue: when first starting the UI, adding cards to the canvas doesn't work, with no explanation why. I think it's because no view is selected by default. Some ways to solve this:
> 
>     * Select the first view in the list at startup
> 
>     * Allow the user to set a "default" view, and select that one at startup
> 
>     * Warn the user that they need to select a view
> 
> 
> The easiest is probably the warning. @anasstarfaoui do you see a better way of dealing with this?

An error toast will now pop if a user tries to double click or drag an item if no view is selected.

**Review by augustebaum — APPROVED** [review]

**Pull request merged**